cmd/minikube/cmd/config/profile_list.go

```diff
@@ -205,7 +205,7 @@ func printProfilesJSON() {
 	updateProfilesStatus(validProfiles)
 
 	var body = map[string]interface{}{}
-	if err == nil || config.IsNotExist(err) {
+	if err == nil || os.IsNotExist(err) {
```

Comment:
This PR is  bypassing the config. package, the better approach would be moving this logic down into the config package pkg/minikube/config/config.go
the relevant code is here:
```
func (c *simpleConfigLoader) LoadConfigFromFile(profileName string, miniHome ...string) (*ClusterConfig, error) {
	var cc ClusterConfig
	// Move to profile package
	path := profileFilePath(profileName, miniHome...)

	if _, err := os.Stat(path); err != nil {
		if os.IsNotExist(err) {
			return nil, &ErrNotExist{fmt.Sprintf("cluster %q does not exist", profileName)}
		}
		return nil, errors.Wrap(err, "stat")
	}


```

so I am curious can we check (by adding temporarily logs) if our code does NOT detect an error here ? 	`if _, err := os.Stat(path); err != nil {`
https://github.com/medyagh/minikube/blob/4d631bab7e32635a90afbbe6a8677e675be7d69c/pkg/minikube/config/config.go#L210

Reply:
I agree this is not the best fix, but it is the simplest way to fix this regression
safely - basically restoring the original fix from 2019.

In general passing an expected condition (no profiles dir) as an error from `profileDirs()` is the real issue.

```
// profileDirs gets all the folders in the user's profiles folder regardless of valid or invalid config
func profileDirs(miniHome ...string) (dirs []string, err error) {                                                                   
    ...   
    pRootDir := filepath.Join(miniPath, "profiles")
    items, err := os.ReadDir(pRootDir)
    for _, f := range items {
        if f.IsDir() {
            dirs = append(dirs, f.Name())
        }   
    }   
    return dirs, err 
}
```

We can handle it early as expected condition:

```
diff --git a/cmd/minikube/cmd/config/profile_list.go b/cmd/minikube/cmd/config/profile_list.go
index 2cd2f1432..61eff8e25 100644
--- a/cmd/minikube/cmd/config/profile_list.go
+++ b/cmd/minikube/cmd/config/profile_list.go
@@ -203,11 +203,11 @@ func warnInvalidProfiles(invalidProfiles []*config.Profile) {
 func printProfilesJSON() {
        validProfiles, invalidProfiles, err := listProfiles()
        updateProfilesStatus(validProfiles)
 
        var body = map[string]interface{}{}
-       if err == nil || config.IsNotExist(err) {
+       if err == nil {
                body["valid"] = profilesOrDefault(validProfiles)
                body["invalid"] = profilesOrDefault(invalidProfiles)
                jsonString, _ := json.Marshal(body)
                os.Stdout.Write(jsonString)
        } else {
diff --git a/pkg/minikube/config/profile.go b/pkg/minikube/config/profile.go
index d14a8f732..040ac8d53 100644
--- a/pkg/minikube/config/profile.go
+++ b/pkg/minikube/config/profile.go
@@ -299,10 +299,16 @@ func profileDirs(miniHome ...string) (dirs []string, err error) {
        if len(miniHome) > 0 {
                miniPath = miniHome[0]
        }
        pRootDir := filepath.Join(miniPath, "profiles")
        items, err := os.ReadDir(pRootDir)
+
+       // Missing profiles dir is expected.
+       if err != nil && os.IsNotExist(err) {
+               return dirs, nil
+       }
+
        for _, f := range items {
                if f.IsDir() {
                        dirs = append(dirs, f.Name())
                }
        }

```

and then no check is needed when printing the json later.

What do you think?

Reply:
@medyagh what do think about https://github.com/kubernetes/minikube/pull/15594#discussion_r1062956489?

Reply:
I think it is better to not rely on os here and use our package that handles it, and if our packag edo snot handle it, we could improve it to do it

Reply:
I'm not sure how do yo suggest to solve this.

The code for listing profiles directories is documented as:
> // profileDirs gets all the folders in the user's profiles folder regardless of valid or invalid config

So using os.ReadDir() look like the right thing to do. The issue is propagating
os.Error to the caller instead of config.Error.

In https://github.com/kubernetes/minikube/pull/15594#discussion_r1062956489
we handle the expected condition of missing profiles directory so we don't
propagate os.Error to the rest of the code. This avoids the confusing when
check if the error is non existing directory.

Reply:
I think what we can do is create a new error type and the first function that returns the err can create this new error type and the upstream funcs can check against this err type

Reply:
This is possible, but I don't think this condition should be reported as an error.
This is an expected condition when starting minikube in the first time, or after 
a user deletes the .minikube directory.